### ISSUE TITLE
MailboxProcessor to Elmish.Program

### ISSUE THREAD

**Issue by ctaggart** [fable-elmish/elmish#135]
I was looking to switch a server-side app from using a MailboxProcessor to using an Elmish.Program. I thought it would be easier to compose and test. The answer could simply be that an Elmish.Program is not the right fit. The two things I am missing are:

- [ ] How to we do a [PostAndReply](https://msdn.microsoft.com/en-us/visualfsharpdocs/conceptual/mailboxprocessor.postandreply%5B%27msg,%27reply%5D-method-%5Bfsharp%5D)?
- [ ] How to we gracefully exit the program loop?

**Comment by et1975** [comment]
Technically, "reply" mechanism can easily be injected into your update function. Graceful loop termination is way harder and really depends on what constitutes "graceful" in  your scenario.
I wrote an (UI-less) agent based on elmish core some time last year as well, but ended up not using it because I wanted the agent to have stronger processing guarantees than simply dropping the message if I'm not able to process it *at this particular moment*. Also all that wrapping and unwrapping of nested structures get's old real quick when you don't have large natural aggregates that UI typically has ;) In short, it may not be a good fit for you either. Maybe take a look at some [FRP library](/stijnmoreels/Sodium.FSharp)?
I ended up extending [FsShelter](/prolucid/FsShelter) with self-hosting instead, so now I can use the same approach for single-server/constrained device and a large cluster.
